src/shogun/io/ARFFFile.cpp

```diff
@@ -0,0 +1,357 @@
+					ss >> date::parse(*date_pos, t);
+					if (bool(ss))
+					{
+						auto value_timestamp = t.time_since_epoch().count();
+						shogun::get<std::vector<float64_t>>(m_data_vectors[i])
+						    .push_back(value_timestamp);
+					}
+					else
+						SG_SERROR(
+						    "Error parsing date \"%s\" with date format \"%s\" "
+						    "on line %d.\n",
+						    elems[i].c_str(), (*date_pos).c_str(),
+						    m_line_number)
+					++date_pos;
+				}
+				break;
+				case (Attribute::String):
+					shogun::get<std::vector<std::string>>(m_data_vectors[i])
+					    .emplace_back(elems[i]);
+				}
+			}
+			if (i != m_attributes.size())
+				SG_SERROR(
+				    "Unexpected number of values on line %d, expected %d "
+				    "values, "
+				    "but found %d.\n",
+				    m_line_number, m_attributes.size(), i)
+			++m_row_count;
+		}
+	};
+	auto check_data = [this]() {
+		// check X values
+		SG_SDEBUG(
+		    "size: %d, cols: %d, rows: %d", m_data_vectors.size(),
+		    m_data_vectors.size() / m_row_count, m_row_count)
+		if (!m_data_vectors.empty())
+		{
+			auto feature_count = m_data_vectors.size();
+			index_t row_count =
+			    shogun::visit(VectorSizeVisitor{}, m_data_vectors[0]);
+			for (int i = 1; i < feature_count; ++i)
+			{
+				REQUIRE(
+				    shogun::visit(VectorSizeVisitor{}, m_data_vectors[i]) ==
+				        row_count,
+				    "All columns must have the same number of features!\n")
+			}
+		}
+		else
+			return false;
+		return true;
+	};
+	process_chunk(read_data, check_data, true);
+}
+
+std::shared_ptr<CCombinedFeatures> ARFFDeserializer::get_features() const
+{
+	auto result = std::make_shared<CCombinedFeatures>();
+	index_t row_count = shogun::visit(VectorSizeVisitor{}, m_data_vectors[0]);
+	for (int i = 0; i < m_data_vectors.size(); ++i)
+	{
+		Attribute att = m_attributes[i];
+		auto vec = m_data_vectors[i];
+		switch (att)
+		{
+		case Attribute::Numeric:
+		case Attribute::Integer:
+		case Attribute::Real:
+		case Attribute::Date:
+		case Attribute::Nominal:
+		{
+			auto casted_vec = shogun::get<std::vector<float64_t>>(vec);
+			SGMatrix<float64_t> mat(1, row_count);
+			memcpy(
+			    mat.matrix, casted_vec.data(),
+			    casted_vec.size() * sizeof(float64_t));
+			auto* feat = new CDenseFeatures<float64_t>(mat);
+			result->append_feature_obj(feat);
+		}
+		break;
+		case Attribute::String:
+		{
+			auto casted_vec = shogun::get<std::vector<std::string>>(vec);
+			index_t max_string_length = 0;
+			for (const auto& el : casted_vec)
+			{
+				if (max_string_length < el.size())
+					max_string_length = el.size();
+			}
+			SGStringList<char> strings(row_count, max_string_length);
+			for (int j = 0; j < row_count; ++j)
+			{
+				SGString<char> current(max_string_length);
+				memcpy(
+				    current.string, casted_vec[j].data(),
+				    (casted_vec.size() + 1) * sizeof(char));
+				strings.strings[j] = current;
+			}
+			auto* feat = new CStringFeatures<char>(strings, EAlphabet::RAWBYTE);
```

Comment:
nice that this works for strings.... what about sparse data?

Reply:
ok as discussed can easily be added later

Reply:
oh btw same thing here with the alphabet.... there might be multiple desired options ... Shogun can also deal with strings of uint16_t etc